cluster/assemblestate/selector.go

```diff
@@ -0,0 +1,380 @@
+package assemblestate
+
+import (
+	"errors"
+	"sort"
+	"time"
+
+	"github.com/snapcore/snapd/cluster/assemblestate/bimap"
+	"github.com/snapcore/snapd/cluster/assemblestate/bitset"
+	"github.com/snapcore/snapd/randutil"
+)
+
+// RouteSelector keeps track of which routes we've seen and helps pick which
+// peers to publish routes to.
+type RouteSelector interface {
+	// AddAuthoritativeRoute records a route from this local node to the given
+	// [RDT]. This route will be published to the given peer, regardless of our
+	// knowledge of that peer's identity.
+	AddAuthoritativeRoute(from DeviceToken, via string)
+
+	// RecordRoutes records a set of routes from the given [RDT]. The given
+	// function identified is used to determine which new routes can be
+	// considered for publication.
+	RecordRoutes(from DeviceToken, r Routes, identified func(DeviceToken) bool) (int, int, error)
+
+	// VerifyRoutes re-calculates which routes are available for publication.
+	// The given function identified is used to determine which routes should be
+	// considered.
```

Comment:
this probably should be less abstract in the description of what happens. Verifying a route means we have identified the peer and that what's identified is for?